UPGRADING.md

```diff
@@ -0,0 +1,207 @@
+duplicate_relation = relation.copy()
+
+# Scenario 7.x
+new_container = dataclasses.replace(container, can_connect=True)
+duplicate_relation = copy.deepcopy(relation)
+```
+
+### Mocking Container.exec
+
+In addition to the new ability to assert against content the charm writes to
+`stdin` and to use pattern matching for the command, `Container.exec_mocks` is
+now named `Container.execs`, `Container.service_status` is now named
+`Container.service_statuses`, and `ExecOutput` is now named `Exec`.
+
+### Resources are classes, not dictionaries
+
+The resources in State objects are now `scenario.Resource` objects rather than
+plain dictionaries, aligning with all of the other State components.
+
+```python
+# Older Scenario code
+state = State(resources={"/path/to/foo", pathlib.Path("/mock/foo")})
+
+# Scenario 7.x
+resource = Resource(location="/path/to/foo", source=pathlib.Path("/mock/foo"))
+state = State(resources={resource})
+```
+
+### State.storage and State.stored_state plurals
+
+The `State.storage` and `State.stored_state` attributes are now plurals, to
+reflect that you may have more than one in the state, aligning with the other
+State components.
+
+```python
+# Older Scenario code
+state = State(stored_state=[ss1, ss2], storage=[s1, s2])
+
+# Scenario 7.x
+state = State(stored_states={s1, s2}, storages={s1, s2})
+```
+
+### Network
+
+`Network` objects are added to the state as a set, like the other components,
+rather than as a dictionary of `{binding_name: network}` as previously. This
+means that the `Network` object now requires a binding name to be passed in when
+it is created.
+
+```python
+# Older Scenario code
+state = State(networks={"foo": Network.default()})
+
+# Scenario 7.x
+state = State(networks={Network.default("foo")})
+```
+
+### Names made private
+
+Several attributes and classes that were never intended for end users have been made private:
+
+* The `data_type_name` attribute of `StoredState` is now private.
+* The `RelationBase` class is now private.
+* The `Event` class is now private.
+
+### Minor removed functionality
+
+The `scenario.sequences` module has been removed. We encourage you to look at
+the new [Catan](https://github.com/PietroPasotti/catan) package.
+
+The `State.jsonpatch_delta()` and `state.sort_patch()` methods have been
+removed. We are considering adding delta-comparisons of state again in the
+future, but have not yet decided how this will look. In the meantime, you can
+use the jsonpatch package directly if necessary. See the tests/helpers.py file
+for an example.
+
+The `Context.cleanup()` and `Context.clear()` methods have been removed. You
+do not need to manually call any cleanup methods after running an event. If you
+want a fresh `Context` (e.g. with no history), you should create a new object.
+
+The deprecated `pre_event` and `post_event` arguments to `run` and `run_action`
+have been removed. Use the appropriate context handler instead.
```

Comment:
Can we specify what the "appropriate context handler" is?

Reply:
I'll make a note to add that, but we have to [make a decision](https://github.com/canonical/ops-scenario/pull/115) first :smile:.